test/llvmcall.jl

```diff
@@ -205,6 +205,34 @@ module CcallableRetTypeTest
     @test do_the_call() === 42.0
 end
 
+# Issue #48093 - test that non-external globals are not deduplicated
+@inline shmem() = Base.llvmcall(("""
+        @shmem = internal global [1 x i8] zeroinitializer, align 32
+
+        define i8* @entry() #0 {
+            ret i8* getelementptr inbounds ([1 x i8], [1 x i8]* @shmem, i64 0, i64 0)
+        }
+
+        attributes #0 = { alwaysinline }""", "entry"),
+    Core.LLVMPtr{Int8,0}, Tuple{})
+
+function kernel2()
+    ptr1 = reinterpret(Ptr{Int8}, shmem())
+    arr1 = unsafe_wrap(Array, ptr1, 1)
+    ptr2 = reinterpret(Ptr{Int8}, shmem())
+    arr2 = unsafe_wrap(Array, ptr2, 1)
+    @inbounds begin
+        arr1[] = 1
+        arr2[]
+    end
+end
+
+get_module_ir(io, f, types) = code_llvm(io, f, types, dump_module=true, raw=true)
+
+ir = sprint(get_module_ir, kernel2, Tuple{})
+
+@test length(collect(eachmatch(r"internal global", ir))) == 2
```

Comment:
It is not legal to assume that the llvmcall content will necessarily be inlined. However, if you make a kernel like this:
```julia
function kernel()
    Base.llvmcall(("""
        @shmem = internal global i8 0, align 8
        define void @entry() {
            store i8 1, i8* @shmem
            ret void
        }""", "entry"), Cvoid, Tuple{})
    Base.llvmcall(("""
        @shmem = internal global i8 0, align 8
        define i8 @entry() {
            %1 = load i8, i8* @shmem
            ret i8 %1
        }""", "entry"), UInt8, Tuple{})
end
```
Then you can make a functional test for this and show it is broken:
```
@test_broken kernel() === 0x00
```